packages/drivers/odsp-driver/src/odspDeltaStorageService.ts

```diff
@@ -194,12 +194,22 @@ export class OdspDeltaStorageWithCache implements IDocumentDeltaStorageService {
 				validateMessages("cached", messagesFromCache, from, this.logger);
 				if (messagesFromCache.length !== 0) {
 					opsFromCache += messagesFromCache.length;
+					// Set the firstCacheMiss at seq number = (fetched ops last seq number) in case we didn't get
+					// all the ops. This will save an extra cache read on "DocumentOpen" and 1 cache read on
+					// "PostDocumentOpen".
+					if (from + messagesFromCache.length < to) {
+						this.firstCacheMiss = Math.min(
+							this.firstCacheMiss,
+							from + messagesFromCache.length - 1,
+						);
+					}
 					return {
 						messages: messagesFromCache,
 						partialResult: true,
 					};
 				}
-				this.firstCacheMiss = Math.min(this.firstCacheMiss, from);
+				// Set at from - 1 to avoid an extra cache read on "PostDocumentOpen".
+				this.firstCacheMiss = Math.min(this.firstCacheMiss, from - 1);
```

Comment:
Looking at the code, this.firstCacheMiss is equal to last op successfully read from cache.
But condition on line 192 assumes that it's first op that was not present in cache (i.e. +1 from above).

I think the old code was right. But what your change does is over-compensate for DeltaManager's -1 game.
I'd rather keep this code honest (i.e. keep old behavior). If we do not do it right, someone (not realizing DM's dealings) will correct this code, and thus making perf worse.

The way I'd make it not do extra cache reads is instead of setting this.firstCacheMiss to max, set it to -1 when we have a cache miss. That way any future reads will not hit cache any more.

Reply:
Improved the logic.